README
--------------------------------------------------------------------------------

INSTALLATION:
-------------
Installation instructions are contained within INSTALL

SUPPORT:
--------
There is a mailing list for discussion of ratbox issues,
 To subscribe, visit:
      http://lists.ratbox.org/cgi-bin/mailman/listinfo/ircd-ratbox

You may also (if youre lucky) find us on #ratbox@EFNet

Note you should *always* check the logfile first.

BUG REPORTS:
------------
If you run this code and encounter problems, you must report
the bug by EMAIL to ircd-ratbox@lists.ratbox.org

Please include a gdb backtrace and keep the core file, binaries and 
modules in case the developers need them.

MODULE RESTARTS:
----------------
Note a running module should never be modified directly whilst it is loaded
into ircd as the behaviour of the OS can result in a core when this is done.

When upgrading modules, you should never simply cp over a module that is
loaded into ircd.  You may delete it then cp, mv it, unload the module from
ircd first, move the directory out of the way and start again or numerous
other things.

Just *dont* do something like: cp m_foo.so ~/ircd/modules/

--------------------------------------------------------------------------------
$Id: README 21191 2005-11-13 22:16:05Z leeh $


ChangeLog
androsyn    2010/01/25 16:41:18 UTC	(20100125_1-26748)

  Log: Update change log
	tag directircd 0.0.1
	Modified:

  Log: Add more channel mode prefixes and support for ~%& modes
  tag directircd 0.0.1
	Modified:
		DirectIRCd/modules/core/m_mode.c
		DirectIRCd/include/channel.h
		DirectIRCd/include/supported.h

  Log: Fork
  tag directircd 0.0.1 
	Branch from ircd-ratbox. HERE WE GO!

  Log:
  tag ratbox 2.2.9

  Modified:
	ircd-ratbox/tags/RATBOX_2_2_9/ (File Added) 


androsyn    2010/01/25 16:36:01 UTC	(20100125_0-26746)
  Log:
  add critical fixes for bugs in /quote HELP and /LINKS - thanks to spb@efnet for pointing these out. Also update RELNOTES and patchlevel.h for 2.2.9

  Modified:
	ircd-ratbox/branches/RATBOX_2_2/RELNOTES (File Modified) 
	ircd-ratbox/branches/RATBOX_2_2/include/patchlevel.h (File Modified) 
	ircd-ratbox/branches/RATBOX_2_2/modules/m_links.c (File Modified) 
	ircd-ratbox/branches/RATBOX_2_2/src/cache.c (File Modified) 
	ircd-ratbox/branches/RATBOX_2_2/src/irc_string.c (File Modified) 
	ircd-ratbox/branches/RATBOX_2_2/src/lex.yy.c (File Modified) 
	ircd-ratbox/branches/RATBOX_2_2/src/y.tab.c (File Modified) 
	ircd-ratbox/branches/RATBOX_2_2/src/y.tab.h (File Modified) 


androsyn    2008/11/22 16:35:58 UTC	(20081122_0-26234)
  Log:
  sid should be defined as sid[4] not [3], somebody forgot space for the trailing \0

  Modified:
	ircd-ratbox/branches/RATBOX_2_2/include/s_conf.h (File Modified) 


androsyn    2008/09/06 15:07:39 UTC	(20080906_0-26024)
  Log:
  add a missing argument to LOAD2HI

  Modified:
	ircd-ratbox/branches/RATBOX_2_2/contrib/m_mkpasswd.c (File Modified) 


androsyn    2008/09/02 19:45:29 UTC	(20080902_0-25988)
  Log:
  add the majority_gline fix to 2.2

  Modified:
	ircd-ratbox/branches/RATBOX_2_2/modules/m_gline.c (File Modified) 


leeh        2008/06/09 19:13:32 UTC	(20080609_0-25526)
  Log:
  - it's impossible to not have an info field in ms_server(), so remove the
    check to shut gcc up
  

  Modified:
	ircd-ratbox/branches/RATBOX_2_2/modules/core/m_server.c (File Modified) 


androsyn    2008/04/21 19:57:03 UTC	(20080421_0-25253)
  Log:
  backport from trunk - don't attempt to fflush a file we just closed

  Modified:
	ircd-ratbox/branches/RATBOX_2_2/src/s_log.c (File Modified) 


jilles      2008/02/07 20:45:23 UTC	(20080207_0-25060)
  Log:
  Fix a UID leak.
  (trunk r25058)
  

  Modified:
	ircd-ratbox/branches/RATBOX_2_2/modules/m_invite.c (File Modified) 


androsyn    2008/01/11 20:28:57 UTC	(20080111_6-24914)
  Log:
  ugh again

  Modified:
	ircd-ratbox/branches/RATBOX_2_2/RELNOTES (File Modified) 
	ircd-ratbox/branches/RATBOX_2_2/contrib/ircd-shortcut.pl (File Added) 
	ircd-ratbox/branches/RATBOX_2_2/include/patchlevel.h (File Modified) 
	ircd-ratbox/branches/RATBOX_2_2/include/s_newconf.h (File Modified) 


androsyn    2008/01/11 20:23:04 UTC	(20080111_5-24912)
  Log:
  add missing prototype

  Modified:
	ircd-ratbox/branches/RATBOX_2_2/include/s_newconf.h (File Modified) 


androsyn    2008/01/11 20:05:58 UTC	(20080111_4-24908)
  Log:
  bleh..update for 2.2.7.1 :(

  Modified:
	ircd-ratbox/branches/RATBOX_2_2/RELNOTES (File Modified) 
	ircd-ratbox/branches/RATBOX_2_2/include/patchlevel.h (File Modified) 


androsyn    2008/01/11 20:04:29 UTC	(20080111_3-24906)
  Log:
  son of a bitch, messed that up

  Modified:
	ircd-ratbox/branches/RATBOX_2_2/src/s_newconf.c (File Modified) 


androsyn    2008/01/11 19:44:20 UTC	(20080111_2-24902)
  Log:
  update RELNOTES and the like for 2.2.7

  Modified:
	ircd-ratbox/branches/RATBOX_2_2/RELNOTES (File Modified) 
	ircd-ratbox/branches/RATBOX_2_2/include/patchlevel.h (File Modified) 
	ircd-ratbox/branches/RATBOX_2_2/src/version.c.SH (File Modified) 


androsyn    2008/01/11 19:02:50 UTC	(20080111_1-24900)
  Log:
  backport jilles match_esc fixes and make s_newconf.c compile again

  Modified:
	ircd-ratbox/branches/RATBOX_2_2/src/match.c (File Modified) 
	ircd-ratbox/branches/RATBOX_2_2/src/s_newconf.c (File Modified) 


androsyn    2008/01/11 18:59:30 UTC	(20080111_0-24898)
  Log:
  Backport find_xline_mask() and find_resv_mask() functions that resolve issues with duplicate
  xline and resv lookups.
  

  Modified:
	ircd-ratbox/branches/RATBOX_2_2/include/s_newconf.h (File Modified) 
	ircd-ratbox/branches/RATBOX_2_2/modules/m_resv.c (File Modified) 
	ircd-ratbox/branches/RATBOX_2_2/modules/m_xline.c (File Modified) 
	ircd-ratbox/branches/RATBOX_2_2/src/s_newconf.c (File Modified) 


androsyn    2007/12/20 18:25:10 UTC	(20071220_0-24754)
  Log:
  Fix a really stupid off by one error when using linebuf_get in raw mode.  Also add the \0 regardless if the buffer is raw or not.
  This bug has been hanging around for about 6 years.  Note that this bug is impossible to trigger in 2.2.x, just fixing it for the sake of sanity.
  
  

  Modified:
	ircd-ratbox/branches/RATBOX_2_2/src/linebuf.c (File Modified) 


jilles      2007/11/07 17:40:44 UTC	(20071107_0-24473)
  Log:
  Backport trunk r23575:
  Include real hostname in Closing Link message for unknown connections
  that have sent NICK or USER, in particular banned users.
  

  Modified:
	ircd-ratbox/branches/RATBOX_2_2/src/client.c (File Modified) 


jilles      2007/09/19 19:25:29 UTC	(20070919_0-24410)
  Log:
  Backport trunk r24408:
  Don't say that services cannot be killed in ERR_ISCHANSERVICE
  message (given when trying to kick or deop them), as that's
  not the case.
  

  Modified:
	ircd-ratbox/branches/RATBOX_2_2/src/messages.tab (File Modified) 


androsyn    2007/07/30 19:49:41 UTC	(20070730_1-24133)
  Log:
  Fix parser to only take a list of values when requested

  Modified:
	ircd-ratbox/branches/RATBOX_2_2/src/ircd_parser.y (File Modified) 
	ircd-ratbox/branches/RATBOX_2_2/src/lex.yy.c (File Modified) 
	ircd-ratbox/branches/RATBOX_2_2/src/y.tab.c (File Modified) 
	ircd-ratbox/branches/RATBOX_2_2/src/y.tab.h (File Modified) 


androsyn    2007/07/30 15:02:14 UTC	(20070730_0-24125)
  Log:

src/md5.c
/*
 * This is an OpenSSL-compatible implementation of the RSA Data Security,
 * Inc. MD5 Message-Digest Algorithm.
 *
 * Written by Solar Designer <solar@openwall.com> in 2001, and placed in
 * the public domain.  There's absolutely no warranty.
 *
 * This differs from Colin Plumb's older public domain implementation in
 * that no 32-bit integer data type is required, there's no compile-time
 * endianness configuration, and the function prototypes match OpenSSL's.
 * The primary goals are portability and ease of use.
 *
 * This implementation is meant to be fast, but not as fast as possible.
 * Some known optimizations are not included to reduce source code size
 * and avoid compile-time configuration.
 */

#include "setup.h"
#if !defined(HAVE_LIBCRYPTO)

#include <string.h>

#include "md5.h"

/*
 * The basic MD5 functions.
 *
 * F is optimized compared to its RFC 1321 definition just like in Colin
 * Plumb's implementation.
 */
#define F(x, y, z)			((z) ^ ((x) & ((y) ^ (z))))
#define G(x, y, z)			((y) ^ ((z) & ((x) ^ (y))))
#define H(x, y, z)			((x) ^ (y) ^ (z))
#define I(x, y, z)			((y) ^ ((x) | ~(z)))

/*
 * The MD5 transformation for all four rounds.
 */
#define STEP(f, a, b, c, d, x, t, s) \
	(a) += f((b), (c), (d)) + (x) + (t); \
	(a) = (((a) << (s)) | (((a) & 0xffffffff) >> (32 - (s)))); \
	(a) += (b);

/*
 * SET reads 4 input bytes in little-endian byte order and stores them
 * in a properly aligned word in host byte order.
 *
 * The check for little-endian architectures which tolerate unaligned
 * memory accesses is just an optimization.  Nothing will break if it
 * doesn't work.
 */
#if defined(__i386__) || defined(__vax__)
#define SET(n) \
	(*(MD5_u32plus *)&ptr[(n) * 4])
#define GET(n) \
	SET(n)
#else
#define SET(n) \
	(ctx->block[(n)] = \
	(MD5_u32plus)ptr[(n) * 4] | \
	((MD5_u32plus)ptr[(n) * 4 + 1] << 8) | \
	((MD5_u32plus)ptr[(n) * 4 + 2] << 16) | \
	((MD5_u32plus)ptr[(n) * 4 + 3] << 24))
#define GET(n) \
	(ctx->block[(n)])
#endif

/*
 * This processes one or more 64-byte data blocks, but does NOT update
 * the bit counters.  There're no alignment requirements.
 */
static void *
body(MD5_CTX * ctx, void *data, unsigned long size)
{
	unsigned char *ptr;
	MD5_u32plus a, b, c, d;
	MD5_u32plus saved_a, saved_b, saved_c, saved_d;

	ptr = data;

	a = ctx->a;
	b = ctx->b;
	c = ctx->c;
	d = ctx->d;

	do
	{
		saved_a = a;
		saved_b = b;
		saved_c = c;
		saved_d = d;

/* Round 1 */
		STEP(F, a, b, c, d, SET(0), 0xd76aa478, 7)
			STEP(F, d, a, b, c, SET(1), 0xe8c7b756, 12)
			STEP(F, c, d, a, b, SET(2), 0x242070db, 17)
			STEP(F, b, c, d, a, SET(3), 0xc1bdceee, 22)
			STEP(F, a, b, c, d, SET(4), 0xf57c0faf, 7)
			STEP(F, d, a, b, c, SET(5), 0x4787c62a, 12)
			STEP(F, c, d, a, b, SET(6), 0xa8304613, 17)
			STEP(F, b, c, d, a, SET(7), 0xfd469501, 22)
			STEP(F, a, b, c, d, SET(8), 0x698098d8, 7)
			STEP(F, d, a, b, c, SET(9), 0x8b44f7af, 12)
			STEP(F, c, d, a, b, SET(10), 0xffff5bb1, 17)
			STEP(F, b, c, d, a, SET(11), 0x895cd7be, 22)
			STEP(F, a, b, c, d, SET(12), 0x6b901122, 7)
			STEP(F, d, a, b, c, SET(13), 0xfd987193, 12)
			STEP(F, c, d, a, b, SET(14), 0xa679438e, 17)
			STEP(F, b, c, d, a, SET(15), 0x49b40821, 22)
/* Round 2 */
			STEP(G, a, b, c, d, GET(1), 0xf61e2562, 5)
			STEP(G, d, a, b, c, GET(6), 0xc040b340, 9)
			STEP(G, c, d, a, b, GET(11), 0x265e5a51, 14)
			STEP(G, b, c, d, a, GET(0), 0xe9b6c7aa, 20)
			STEP(G, a, b, c, d, GET(5), 0xd62f105d, 5)
			STEP(G, d, a, b, c, GET(10), 0x02441453, 9)
			STEP(G, c, d, a, b, GET(15), 0xd8a1e681, 14)
			STEP(G, b, c, d, a, GET(4), 0xe7d3fbc8, 20)
			STEP(G, a, b, c, d, GET(9), 0x21e1cde6, 5)
			STEP(G, d, a, b, c, GET(14), 0xc33707d6, 9)
			STEP(G, c, d, a, b, GET(3), 0xf4d50d87, 14)
			STEP(G, b, c, d, a, GET(8), 0x455a14ed, 20)
			STEP(G, a, b, c, d, GET(13), 0xa9e3e905, 5)
			STEP(G, d, a, b, c, GET(2), 0xfcefa3f8, 9)
			STEP(G, c, d, a, b, GET(7), 0x676f02d9, 14)
			STEP(G, b, c, d, a, GET(12), 0x8d2a4c8a, 20)
/* Round 3 */
			STEP(H, a, b, c, d, GET(5), 0xfffa3942, 4)
			STEP(H, d, a, b, c, GET(8), 0x8771f681, 11)
			STEP(H, c, d, a, b, GET(11), 0x6d9d6122, 16)
			STEP(H, b, c, d, a, GET(14), 0xfde5380c, 23)
			STEP(H, a, b, c, d, GET(1), 0xa4beea44, 4)
			STEP(H, d, a, b, c, GET(4), 0x4bdecfa9, 11)
			STEP(H, c, d, a, b, GET(7), 0xf6bb4b60, 16)
			STEP(H, b, c, d, a, GET(10), 0xbebfbc70, 23)
			STEP(H, a, b, c, d, GET(13), 0x289b7ec6, 4)
			STEP(H, d, a, b, c, GET(0), 0xeaa127fa, 11)
			STEP(H, c, d, a, b, GET(3), 0xd4ef3085, 16)
			STEP(H, b, c, d, a, GET(6), 0x04881d05, 23)
			STEP(H, a, b, c, d, GET(9), 0xd9d4d039, 4)
			STEP(H, d, a, b, c, GET(12), 0xe6db99e5, 11)
			STEP(H, c, d, a, b, GET(15), 0x1fa27cf8, 16)
			STEP(H, b, c, d, a, GET(2), 0xc4ac5665, 23)
/* Round 4 */
			STEP(I, a, b, c, d, GET(0), 0xf4292244, 6)
			STEP(I, d, a, b, c, GET(7), 0x432aff97, 10)
			STEP(I, c, d, a, b, GET(14), 0xab9423a7, 15)
			STEP(I, b, c, d, a, GET(5), 0xfc93a039, 21)
			STEP(I, a, b, c, d, GET(12), 0x655b59c3, 6)
			STEP(I, d, a, b, c, GET(3), 0x8f0ccc92, 10)
			STEP(I, c, d, a, b, GET(10), 0xffeff47d, 15)
			STEP(I, b, c, d, a, GET(1), 0x85845dd1, 21)
			STEP(I, a, b, c, d, GET(8), 0x6fa87e4f, 6)
			STEP(I, d, a, b, c, GET(15), 0xfe2ce6e0, 10)
			STEP(I, c, d, a, b, GET(6), 0xa3014314, 15)
			STEP(I, b, c, d, a, GET(13), 0x4e0811a1, 21)
			STEP(I, a, b, c, d, GET(4), 0xf7537e82, 6)
			STEP(I, d, a, b, c, GET(11), 0xbd3af235, 10)
			STEP(I, c, d, a, b, GET(2), 0x2ad7d2bb, 15)
			STEP(I, b, c, d, a, GET(9), 0xeb86d391, 21) a += saved_a;
		b += saved_b;
		c += saved_c;
		d += saved_d;

		ptr += 64;
	}
	while(size -= 64);

	ctx->a = a;
	ctx->b = b;
	ctx->c = c;
	ctx->d = d;

	return ptr;
}

void
MD5_Init(MD5_CTX * ctx)
{
	ctx->a = 0x67452301;
	ctx->b = 0xefcdab89;
	ctx->c = 0x98badcfe;
	ctx->d = 0x10325476;

	ctx->lo = 0;
	ctx->hi = 0;
}

void
MD5_Update(MD5_CTX * ctx, void *data, unsigned long size)
{
	MD5_u32plus saved_lo;
	unsigned long used, free;

	saved_lo = ctx->lo;
	if((ctx->lo = (saved_lo + size) & 0x1fffffff) < saved_lo)
		ctx->hi++;
	ctx->hi += size >> 29;

	used = saved_lo & 0x3f;

	if(used)
	{
		free = 64 - used;

		if(size < free)
		{
			memcpy(&ctx->buffer[used], data, size);
			return;
		}

		memcpy(&ctx->buffer[used], data, free);
		data = (unsigned char *) data + free;
		size -= free;
		body(ctx, ctx->buffer, 64);
	}

	if(size >= 64)
	{
		data = body(ctx, data, size & ~(unsigned long) 0x3f);
		size &= 0x3f;
	}

	memcpy(ctx->buffer, data, size);
}

void
MD5_Final(unsigned char *result, MD5_CTX * ctx)
{
	unsigned long used, free;

	used = ctx->lo & 0x3f;

	ctx->buffer[used++] = 0x80;

	free = 64 - used;

	if(free < 8)
	{
		memset(&ctx->buffer[used], 0, free);
		body(ctx, ctx->buffer, 64);
		used = 0;
		free = 64;
	}

	memset(&ctx->buffer[used], 0, free - 8);

	ctx->lo <<= 3;
	ctx->buffer[56] = ctx->lo;
	ctx->buffer[57] = ctx->lo >> 8;
	ctx->buffer[58] = ctx->lo >> 16;
	ctx->buffer[59] = ctx->lo >> 24;
	ctx->buffer[60] = ctx->hi;
	ctx->buffer[61] = ctx->hi >> 8;
	ctx->buffer[62] = ctx->hi >> 16;
	ctx->buffer[63] = ctx->hi >> 24;

	body(ctx, ctx->buffer, 64);

	result[0] = ctx->a;
	result[1] = ctx->a >> 8;
	result[2] = ctx->a >> 16;
	result[3] = ctx->a >> 24;
	result[4] = ctx->b;
	result[5] = ctx->b >> 8;
	result[6] = ctx->b >> 16;
	result[7] = ctx->b >> 24;
	result[8] = ctx->c;
	result[9] = ctx->c >> 8;
	result[10] = ctx->c >> 16;
	result[11] = ctx->c >> 24;
	result[12] = ctx->d;
	result[13] = ctx->d >> 8;
	result[14] = ctx->d >> 16;
	result[15] = ctx->d >> 24;

	memset(ctx, 0, sizeof(ctx));
}

#else
/* need these includes for DoMD5()... */
#include "stdinc.h"
#include "common.h"
#endif

/** Generates an MD5 checksum.
 * @param mdout[out] Buffer to store result in, the result will be 16 bytes in binary
 *                   (not ascii printable!).
 * @param src[in]    The input data used to generate the checksum.
 * @param n[in]      Length of data.
 */
void
DoMD5(unsigned char *mdout, unsigned char *src, unsigned long n)
{
	MD5_CTX hash;

	MD5_Init(&hash);
	MD5_Update(&hash, src, n);
	MD5_Final(mdout, &hash);
}
